### PR TITLE
fix for nxos_ospf_vrf invalid cmd and ntp errors

### LINKED ISSUE #27392: (P1) nxos_ospf_vrf module sends invalid command
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - nxos_ospf_vrf

##### NXOS VERSION
7.0(3)I5(3)

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel b93b9e68d7) last updated 2017/07/26 14:52:17 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible']
  ansible python module location = /root/agents-ci/ansible/lib/ansible
  executable location = /root/agents-ci/ansible/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```

##### OS / ENVIRONMENT
Ubuntu 14.04

##### SUMMARY
The `nxos_ospf_vrf` module sends an invalid command using the following playbook.  Problem exists for both `nxapi` and `cli`
```yaml
---
- debug: msg="START TRANSPORT:{{ connection.transport }} nxos_ospf_vrf sanity test"

- name: "Enable feature OSPF"
  nxos_feature:
    feature: ospf
    state: enabled
    provider: "{{ connection }}"
  ignore_errors: yes

- block:
  - name: Configure ospf vrf
    nxos_ospf_vrf: &config
      ospf: 1
      timer_throttle_spf_start: 50
      timer_throttle_spf_hold: 1000
      timer_throttle_spf_max: 2000
      timer_throttle_lsa_start: 60
      timer_throttle_lsa_hold: 1100
      timer_throttle_lsa_max: 3000
      vrf: test
      state: present
      provider: "{{ connection }}"
    register: result

  - assert: &true
      that:
        - "result.changed == true"

  - name: "Check Idempotence"
    nxos_ospf: *config
    register: result

  - assert: &false
      that:
        - "result.changed == false"

  rescue:
  - name: "Disable feature OSPF"
    nxos_feature:
      feature: ospf
      state: disabled
      provider: "{{ connection }}"
    ignore_errors: yes

  always:
  - name: Unconfigure ospf vrf
    nxos_ospf_vrf: &unconfig
      ospf: 1
      state: absent
      provider: "{{ connection }}"
    register: result

  - assert: *true

  - name: "Check Idempotence"
    nxos_ospf: *unconfig
    register: result

  - assert: *false

  - debug: msg="END TRANSPORT:{{ connection.transport }} nxos_ospf_vrf sanity test"
```

```shell
TASK [nxos_ospf_vrf : debug] ************************************************************************************************************************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_ospf_vrf/tests/common/sanity.yaml:2
ok: [n9k.example.com] => {
    "msg": "START TRANSPORT:cli nxos_ospf_vrf sanity test"
}

TASK [nxos_ospf_vrf : Enable feature OSPF] **********************************************************************************************************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_ospf_vrf/tests/common/sanity.yaml:4
<n9k.example.com> connection transport is cli
<n9k.example.com> using connection plugin network_cli
<n9k.example.com> socket_path: /root/.ansible/pc/4cc10f215f
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/basic.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/nxos.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/_text.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/parsing/__init__.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/pycompat24.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/six/__init__.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/connection.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/network_common.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/urls.py
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_feature.py
<n9k.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n9k.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175 `" && echo ansible-tmp-1501177595.03-269280703013175="` echo $HOME/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175 `" ) && sleep 0'
<n9k.example.com> PUT /tmp/tmpIr4cde TO /root/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175/nxos_feature.py
<n9k.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175/ /root/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175/nxos_feature.py && sleep 0'
<n9k.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175/nxos_feature.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1501177595.03-269280703013175/" > /dev/null 2>&1 && sleep 0'
changed: [n9k.example.com] => {
    "changed": true, 
    "commands": [
        "feature ospf"
    ], 
    "failed": false, 
    "invocation": {
        "module_args": {
            "config": null, 
            "feature": "ospf", 
            "host": null, 
            "include_defaults": null, 
            "password": null, 
            "port": null, 
            "provider": null, 
            "save": null, 
            "ssh_keyfile": null, 
            "state": "enabled", 
            "timeout": 10, 
            "transport": "cli", 
            "use_ssl": null, 
            "username": null, 
            "validate_certs": null
        }
    }
}

TASK [nxos_ospf_vrf : Configure ospf vrf] ***********************************************************************************************************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_ospf_vrf/tests/common/sanity.yaml:12
<n9k.example.com> connection transport is cli
<n9k.example.com> using connection plugin network_cli
<n9k.example.com> socket_path: /root/.ansible/pc/4cc10f215f
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/basic.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/netcfg.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/nxos.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/_text.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/parsing/__init__.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/pycompat24.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/six/__init__.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/network_common.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/connection.py
Using module_utils file /root/agents-ci/ansible/lib/ansible/module_utils/urls.py
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_ospf_vrf.py
<n9k.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n9k.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449 `" && echo ansible-tmp-1501177597.52-53233907747449="` echo $HOME/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449 `" ) && sleep 0'
<n9k.example.com> PUT /tmp/tmpL1tkiO TO /root/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449/nxos_ospf_vrf.py
<n9k.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449/ /root/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449/nxos_ospf_vrf.py && sleep 0'
<n9k.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449/nxos_ospf_vrf.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1501177597.52-53233907747449/" > /dev/null 2>&1 && sleep 0'
fatal: [n9k.example.com]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "auto_cost": null, 
            "config": null, 
            "default_metric": null, 
            "host": null, 
            "include_defaults": "True", 
            "log_adjacency": null, 
            "ospf": "1", 
            "password": null, 
            "port": null, 
            "provider": null, 
            "router_id": null, 
            "save": false, 
            "ssh_keyfile": null, 
            "state": "present", 
            "timeout": 10, 
            "timer_throttle_lsa_hold": "1100", 
            "timer_throttle_lsa_max": "3000", 
            "timer_throttle_lsa_start": "60", 
            "timer_throttle_spf_hold": "1000", 
            "timer_throttle_spf_max": "2000", 
            "timer_throttle_spf_start": "50", 
            "transport": "cli", 
            "use_ssl": null, 
            "username": null, 
            "validate_certs": null, 
            "vrf": "test"
        }
    }, 
    "msg": "ospf 1\r\r\n                             ^\r\n% Invalid command at '^' marker.\r\n\rn9k-109(config-router-vrf)# "
}

```

### PR DESCRIPTION
**Pull request by saichint**
##### SUMMARY
Fixes 
#27392
#27708

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
nxos_ospf_vrf
nxos_ntp

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel bb5b1680f8) last updated 2017/07/26 08:18:32 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/ansible']
  ansible python module location = /root/agents-ci/ansible/lib/ansible
  executable location = /root/agents-ci/ansible/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```

##### ADDITIONAL INFORMATION
- There is a bug in the code where it is taking ospf_instance_id from the playbook and trying to execute a command "ospf <ospf_instance_id>" under the vrf context (if vrf is non-default or under 'router ospf <ospf_instance_id>' if vrf is default). There is no such command. The ospf_instance_id from the playbook must only used to go to the "router ospf <ospf_instance_id>" context. All other commands are fine.

- show run | i ntp.source command returns [''] when the configuration does not exist and this is not handled properly.

- for nxos_ntp {{ nxapi }} case, the piping issue is not handled.

- Added integration test cases

```

```

### CODE REVIEW

**Comment by jborean93** [comment]
Hey @saichint, thanks for the PR, I would probably recommend splitting this so there is 1 PR per module that you touch. If they need to be together that is fine but it is usually easier and quicker to look at smaller chunks.

**Comment by saichint** [comment]
Can we do this one time. Next time I will open separate PRs.
The code changes are small. The number of files are more as they are integration test cases.


On Aug 9, 2017 7:36 PM, Jordan Borean <notifications@github.com> wrote:

Hey @saichint<https://github.com/saichint>, thanks for the PR, I would probably recommend splitting this so there is 1 PR per module that you touch. If they need to be together that is fine but it is usually easier and quicker to look at smaller chunks.

**Comment by jborean93** [comment]
Up to the people reviewing it :), it should be ok though. I don't know much about this domain so I cannot review it myself.

File: lib/ansible/modules/network/nxos/nxos_ntp.py [hunk 1]
```diff
@@ -164,15 +164,16 @@ def get_ntp_source(module):
...
-        try:
-            if 'interface' in output[0]:
-                source_type = 'source-interface'
-            else:
-                source_type = 'source'
-            source = output[0].split()[2].lower()
-        except AttributeError:
-            source_type = None
-            source = None
+        if output[0]:
```
Comment (Qalthos) [inline lib/ansible/modules/network/nxos/nxos_ntp.py]: Take this out and replace the `except` line to `except (AttributeError, TypeError)`

**Review by Qalthos — CHANGES_REQUESTED** [review]
I know it's minor, but this minimizes the diff and tries to avoid runaway nesting.

Other than that, good to merge.

**Comment by saichint** [comment]
@Qalthos I think you meant to change except to except (AttributeError, IndexError). We are not getting any TypeError, if I put that, it still fails. I changed it to IndexError and now it is fine. If this is ok, Please merge.

**Pull request merged**